### PR TITLE
Add missing/move/fix default data methods of all entities

### PR DESCRIPTION
**Pull request by Lignium**

### CODE REVIEW

File: src/main/java/org/spongepowered/api/entity/vehicle/minecart/MinecartLike.java [hunk 1]
```diff
@@ -90,4 +85,23 @@
...
+    default Value.Mutable<Integer> minecartBlockOffset() {
+        return this.requireValue(Keys.MINECART_BLOCK_OFFSET).asMutable();
+    }
+
+    /**
+     * {@link Keys#BLOCK_STATE}
+     *
+     * @return The block state of the block occupied minecart
+     */
+    default Optional<Value.Mutable<BlockState>> block() {
```
Comment (Zidane) [inline src/main/java/org/spongepowered/api/entity/vehicle/minecart/MinecartLike.java]: Why was this moved up to `MinecartLike`?

File: src/main/java/org/spongepowered/api/entity/projectile/explosive/FireworkRocket.java [hunk 2]
```diff
@@ -45,4 +50,12 @@
         return this.requireValue(Keys.FIREWORK_EFFECTS).asMutable();
     }
 
+    /**
+     * {@link Keys#FIREWORK_FLIGHT_MODIFIER}
+     *
+     * @return The flight duration of the firework rocket
+     */
+    default Optional<Value.Mutable<Ticks>> fireworkFlightModifier() {
```
Comment (Zidane) [inline src/main/java/org/spongepowered/api/entity/projectile/explosive/FireworkRocket.java]: `flightModifier`

File: src/main/java/org/spongepowered/api/entity/vehicle/minecart/MinecartLike.java [hunk 3]
```diff
@@ -90,4 +85,23 @@
     default Value.Mutable<Vector3d> derailedVelocityModifier() {
         return this.requireValue(Keys.DERAILED_VELOCITY_MODIFIER).asMutable();
     }
+
+    /**
+     * {@link Keys#MINECART_BLOCK_OFFSET}
+     *
+     * @return The represented block's offset of the minecart
+     */
+    default Value.Mutable<Integer> minecartBlockOffset() {
```
Comment (Zidane) [inline src/main/java/org/spongepowered/api/entity/vehicle/minecart/MinecartLike.java]: `blockOffset` but should be back in `BlockOccupiedMinecart`

File: src/main/java/org/spongepowered/api/entity/AreaEffectCloud.java [hunk 4]
```diff
@@ -128,14 +128,4 @@
     default ListValue.Mutable<PotionEffect> effects() {
         return this.requireValue(Keys.POTION_EFFECTS).asMutable();
     }
-
-    /**
-     * {@link Keys#AGE}
-     *
-     * @return The age of this cloud
-     */
-    default Value.Mutable<Integer> age() {
```
Comment (gabizou) [inline src/main/java/org/spongepowered/api/entity/AreaEffectCloud.java:137]: IIRC this was included to basically exist as a way to determine how much of the cloud remains `if age > duration + waitTime) { setDead(); }`

File: src/main/java/org/spongepowered/api/entity/living/player/Player.java [hunk 5]
```diff
@@ -61,4 +67,194 @@ default GameProfile identity() {
...
+    default Value.Mutable<Integer> sleepTimer() {
+        return this.requireValue(Keys.SLEEP_TIMER).asMutable();
+    }
+
+    /**
+     * {@link Keys#FIRST_DATE_JOINED}
+     *
+     * @return The timestamp value when this player first joined
+     */
+    default Optional<Value.Mutable<Instant>> firstJoined() {
```
Comment (Zidane) [inline src/main/java/org/spongepowered/api/entity/living/player/Player.java]: This should be on `ServerPlayer`

`Player` can also be a `ClientPlayer` which will not know this information.

File: src/main/java/org/spongepowered/api/entity/living/player/Player.java [hunk 6]
```diff
@@ -61,4 +67,194 @@ default GameProfile identity() {
...
+        return this.getValue(Keys.FIRST_DATE_JOINED).map(Value::asMutable);
+    }
+
+    /**
+     * {@link Keys#LAST_DATE_JOINED}
+     *
+     * @return The last timestamp value when this player has joined
+     */
+    default Optional<Value.Mutable<Instant>> lastJoined() {
```
Comment (Zidane) [inline src/main/java/org/spongepowered/api/entity/living/player/Player.java]: See above comment

File: src/main/java/org/spongepowered/api/entity/living/player/Player.java [hunk 7]
```diff
@@ -61,4 +67,194 @@ default GameProfile identity() {
...
+        return this.getValue(Keys.LAST_DATE_JOINED).map(Value::asMutable);
+    }
+
+    /**
+     * {@link Keys#LAST_DATE_PLAYED}
+     *
+     * @return The last timestamp value when this player has played
+     */
+    default Optional<Value.Mutable<Instant>> lastPlayed() {
```
Comment (Zidane) [inline src/main/java/org/spongepowered/api/entity/living/player/Player.java]: See above comment

File: src/main/java/org/spongepowered/api/entity/living/Humanoid.java [hunk 8]
```diff
@@ -30,44 +30,19 @@
...
     /**
-     * {@link Keys#GAME_MODE}
+     * {@link Keys#SKIN_PROFILE_PROPERTY}
      *
-     * @return The gamemode
+     * @return The skin of the humanoid
      */
-    default Value.Mutable<GameMode> gameMode() {
-        return this.requireValue(Keys.GAME_MODE).asMutable();
+    default Value.Mutable<ProfileProperty> skinProfileProperty() {
```
Comment (Zidane) [inline src/main/java/org/spongepowered/api/entity/living/Humanoid.java]: `skinProfile`

File: src/main/java/org/spongepowered/api/entity/AreaEffectCloud.java (hunk 4 above)
Comment (gabizou) [reply src/main/java/org/spongepowered/api/entity/AreaEffectCloud.java:137]: You could explicitly override to provide some documentation about the age being used

File: src/main/java/org/spongepowered/api/entity/Entity.java [hunk 9]
```diff
@@ -439,19 +440,208 @@ default boolean canSee(final Entity entity) {
...
+    default Value<Double> baseSize() {
+        return this.requireValue(Keys.BASE_SIZE);
+    }
+
+    /**
+     * {@link Keys#EYE_HEIGHT}
+     *
+     * @return The height of the eyes
+     */
+    default Value<Double> eyeHeight() {
```
Comment (Zidane) [inline src/main/java/org/spongepowered/api/entity/Entity.java:481]: May be wrong but I don't think all entities have eye height and position.

File: src/main/java/org/spongepowered/api/entity/living/Humanoid.java [hunk 10]
```diff
@@ -30,44 +30,19 @@
...
 /**
  * Represents a human-like entity in game, such as {@link Player} or {@link Human}s.
  */
 public interface Humanoid extends Living, ArmorEquipable, Tamer {
-    /**
-     * {@link Keys#FOOD_LEVEL}
-     *
-     * @return The human's food level
-     */
-    default Value.Mutable<Integer> foodLevel() {
```
Comment (gabizou) [inline src/main/java/org/spongepowered/api/entity/living/Humanoid.java:43]: do "Human"s not have food and such?

File: src/main/java/org/spongepowered/api/entity/living/Humanoid.java [hunk 11]
```diff
@@ -30,44 +30,19 @@
...
-    }
 
     /**
-     * {@link Keys#GAME_MODE}
+     * {@link Keys#SKIN_PROFILE_PROPERTY}
      *
-     * @return The gamemode
+     * @return The skin of the humanoid
      */
-    default Value.Mutable<GameMode> gameMode() {
```
Comment (gabizou) [inline src/main/java/org/spongepowered/api/entity/living/Humanoid.java:70]: And gamemode? I could've sworn we'd be able to have this part in the entity to expose, or is that only for tab list management?

File: src/main/java/org/spongepowered/api/entity/living/Living.java [hunk 12]
```diff
@@ -102,66 +104,75 @@
...
+    default Optional<Value.Mutable<Boolean>> canGrief() {
+        return this.getValue(Keys.CAN_GRIEF).map(Value::asMutable);
+    }
+
+    /**
+     * {@link Keys#POTION_EFFECTS}
+     *
+     * @return The potion effects that are present on the living entity
+     */
+    default Optional<ListValue.Mutable<PotionEffect>> potionEffects() {
```
Comment (gabizou) [inline src/main/java/org/spongepowered/api/entity/living/Living.java]: Are we considering an empty list to be an empty optional? Or are we saying that not all living entities can have potion effects?

File: src/main/java/org/spongepowered/api/entity/living/animal/Parrot.java [hunk 13]
```diff
@@ -31,7 +31,7 @@
 /**
  * Represents a Parrot.
  */
-public interface Parrot extends TameableAnimal, Sittable {
+public interface Parrot extends TameableAnimal {
```
Comment (gabizou) [inline src/main/java/org/spongepowered/api/entity/living/animal/Parrot.java:34]: Could've sworn parrots can sit?

File: src/main/java/org/spongepowered/api/entity/living/animal/Wolf.java [hunk 14]
```diff
@@ -31,11 +31,12 @@
 /**
  * Represents a Wolf.
  */
-public interface Wolf extends TameableAnimal, Sittable {
+public interface Wolf extends TameableAnimal {
```
Comment (gabizou) [inline src/main/java/org/spongepowered/api/entity/living/animal/Wolf.java:34]: Same with wolves?

File: src/main/java/org/spongepowered/api/entity/vehicle/minecart/MinecartLike.java (hunk 1 above)
Comment (Lignium) [reply src/main/java/org/spongepowered/api/entity/vehicle/minecart/MinecartLike.java]: https://github.com/SpongePowered/Sponge/blob/1d33b3c3daf2fbcee44554d128deec16c11aa96e/src/main/java/org/spongepowered/common/data/provider/entity/AbstractMinecartData.java#L50-L54

File: src/main/java/org/spongepowered/api/entity/vehicle/minecart/MinecartLike.java (hunk 3 above)
Comment (Lignium) [reply src/main/java/org/spongepowered/api/entity/vehicle/minecart/MinecartLike.java]: https://github.com/SpongePowered/Sponge/blob/1d33b3c3daf2fbcee44554d128deec16c11aa96e/src/main/java/org/spongepowered/common/data/provider/entity/AbstractMinecartData.java#L64-L67

File: src/main/java/org/spongepowered/api/entity/living/player/Player.java (hunk 5 above)
Comment (Lignium) [reply src/main/java/org/spongepowered/api/entity/living/player/Player.java]: https://github.com/SpongePowered/Sponge/blob/1d33b3c3daf2fbcee44554d128deec16c11aa96e/src/main/java/org/spongepowered/common/data/provider/entity/IdentifiableData.java#L47-L54

File: src/main/java/org/spongepowered/api/entity/living/player/Player.java (hunk 6 above)
Comment (Lignium) [reply src/main/java/org/spongepowered/api/entity/living/player/Player.java]: https://github.com/SpongePowered/Sponge/blob/1d33b3c3daf2fbcee44554d128deec16c11aa96e/src/main/java/org/spongepowered/common/data/provider/entity/IdentifiableData.java#L55-L62

File: src/main/java/org/spongepowered/api/entity/Entity.java (hunk 9 above)
Comment (Lignium) [reply src/main/java/org/spongepowered/api/entity/Entity.java:481]: https://github.com/SpongePowered/Sponge/blob/1d33b3c3daf2fbcee44554d128deec16c11aa96e/src/main/java/org/spongepowered/common/data/provider/entity/EntityData.java#L78-L81

File: src/main/java/org/spongepowered/api/entity/living/player/Player.java (hunk 5 above)
Comment (Zidane) [reply src/main/java/org/spongepowered/api/entity/living/player/Player.java]: That supports line is wrong, thanks for bringing this to our attention. Should be `h instanceof ServerPlayer`

**Comment by Lignium** [comment]
This PR was done for the most part according to the implementation.

**Comment by Zidane** [comment]
@Lignium Absolutely, you've shown a light at some spots that needed review. Thanks.

File: src/main/java/org/spongepowered/api/entity/vehicle/minecart/MinecartLike.java (hunk 1 above)
Comment (Zidane) [reply src/main/java/org/spongepowered/api/entity/vehicle/minecart/MinecartLike.java]: So this is a strange one in Vanilla:

```java
    public BlockState getDisplayBlockState() {
        return !this.hasCustomDisplay() ? this.getDefaultDisplayBlockState() : Block.stateById((Integer)this.getEntityData().get(DATA_ID_DISPLAY_BLOCK));
    }

    public BlockState getDefaultDisplayBlockState() {
        return Blocks.AIR.defaultBlockState();
    }
```
    
Just why Mojang...

I think we need to ignore that `AbstractMinecart` has this and keep it on our `BlockOccupiedMinecart`

**Comment by dualspiral** [comment]
I don't see anything wrong beyond what you've already fixed. Thanks for the effort here!

**Comment by Lignium** [comment]
There are keys that could not be added due to name collisions with other methods, these are Keys.SCALE and Keys.LOCALE.

File: src/main/java/org/spongepowered/api/entity/Entity.java [hunk 15]
```diff
@@ -369,16 +370,16 @@ default boolean canSee(final Entity entity) {
      *
      * @return The "base vehicle" of the entity vehicle riding chain
      */
-    default Optional<Value.Mutable<Entity>> baseVehicle() {
-        return this.getValue(Keys.BASE_VEHICLE).map(Value::asMutable);
+    default Value<Entity> baseVehicle() {
```
Comment (gabizou) [inline src/main/java/org/spongepowered/api/entity/Entity.java]: Why is this not an optional?

File: src/main/java/org/spongepowered/api/entity/AreaEffectCloud.java (hunk 4 above)
Comment (Zidane) [reply src/main/java/org/spongepowered/api/entity/AreaEffectCloud.java:137]: @gabizou This was put into `Entity`

File: src/main/java/org/spongepowered/api/entity/living/Humanoid.java (hunk 10 above)
Comment (Lignium) [reply src/main/java/org/spongepowered/api/entity/living/Humanoid.java:43]: `HumanEntity` is an artificial class, it is not in the vanilla server. It does not store or process `FoodData`. Only real player has it. However, the pseudo-player does not need this.

File: src/main/java/org/spongepowered/api/entity/living/Humanoid.java (hunk 11 above)
Comment (Lignium) [reply src/main/java/org/spongepowered/api/entity/living/Humanoid.java:70]: The same as with `FoodData`.

File: src/main/java/org/spongepowered/api/entity/living/Living.java (hunk 12 above)
Comment (Lignium) [reply src/main/java/org/spongepowered/api/entity/living/Living.java]: Made according to the implementation: https://github.com/SpongePowered/Sponge/blob/7ae5c4e1c5be2643510a390b195e7ba2fd0d6142/src/main/java/org/spongepowered/common/data/provider/entity/LivingData.java#L153-L164

File: src/main/java/org/spongepowered/api/entity/living/animal/Parrot.java (hunk 13 above)
Comment (Lignium) [reply src/main/java/org/spongepowered/api/entity/living/animal/Parrot.java:34]: All `TameableEntities` can sit, for all of them the game has a flag in the `SynchedEntityData`.

File: src/main/java/org/spongepowered/api/entity/Entity.java (hunk 15 above)
Comment (Lignium) [reply src/main/java/org/spongepowered/api/entity/Entity.java]: From the `net.minecraft.world.entity.Entity`:
```java
public Entity getRootVehicle() {
    Entity var0;
    for(var0 = this; var0.isPassenger(); var0 = var0.getVehicle()) {
    }

    return var0;
}
```
At a minimum, an entity itself is considered the root vehicle.

File: src/main/java/org/spongepowered/api/entity/living/Living.java (hunk 12 above)
Comment (dualspiral) [reply src/main/java/org/spongepowered/api/entity/living/Living.java]: Not a problem with this PR - more that you've highlighted it - but based on the impl you linked I think this shouldn't be optional and the linked impl should not be returning null but an empty list in the scenario of no active effects.

File: src/main/java/org/spongepowered/api/entity/Entity.java (hunk 15 above)
Comment (dualspiral) [reply src/main/java/org/spongepowered/api/entity/Entity.java]: Again, another question that is thrown up because of how you've mirrored it. In minecraft, the base vehicle certainly is treated as one's self if you're not riding anything - a good question here, however, is "does it make sense in the context of the API"?

More to the point, from an API point of view, if an entity is not riding anything, do we want this to return the entity itself, or should it be an empty optional?

File: src/main/java/org/spongepowered/api/entity/living/animal/Parrot.java (hunk 13 above)
Comment (dualspiral) [reply src/main/java/org/spongepowered/api/entity/living/animal/Parrot.java:34]: Yeah, this interfeace extension was moved to `TameableAnimal`, so this removal is fine.

File: src/main/java/org/spongepowered/api/entity/living/Living.java (hunk 12 above)
Comment (Zidane) [reply src/main/java/org/spongepowered/api/entity/living/Living.java]: This should be an empty list in the impl PR

File: src/main/java/org/spongepowered/api/entity/Entity.java (hunk 15 above)
Comment (Zidane) [reply src/main/java/org/spongepowered/api/entity/Entity.java]: This should remain an `Optional`, the fact Vanilla treats yourself as a vehicle when you are not riding anything is an implementation detail.

**Comment by dualspiral** [comment]
https://github.com/SpongePowered/SpongeAPI/commit/a7ee7bd3d7b6426c02d80d2bc3e864f0a019308d
